src/ast/ast.c

```diff
@@ -504,34 +497,50 @@ inline AST_AnnotationCtxCollection *AST_GetAnnotationCtxCollection(AST *ast) {
 	return ast->anot_ctx_collection;
 }
 
-static void _free_to_string_annotation
-(
-	void *userdata,
-	const cypher_astnode_t *node,
-	void *annotation
-) {
-	rm_free(annotation);
+// Compute the number of digits in a non-negative integer.
+static inline int _digit_count(int n) {
+	int count = 0;
+	do {
+		n /= 10;
+		count++;
+	} while(n);
+	return count;
+}
+
+static inline char *_create_anon_alias(int anon_count) {
+	// We need space for "anon_" (5), all digits, and a NULL terminator (1)
+	int alias_len = _digit_count(anon_count) + 6;
+	char *alias = rm_malloc(alias_len * sizeof(char));
+	snprintf(alias, alias_len, "anon_%d", anon_count);
+	return alias;
 }
 
 char *AST_ToString(const cypher_astnode_t *node) {
 	QueryCtx *ctx = QueryCtx_GetQueryCtx();
 	AST *ast = QueryCtx_GetAST();
 	AnnotationCtx *to_string_ctx = AST_AnnotationCtxCollection_GetToStringCtx(ast->anot_ctx_collection);
 
-	if(to_string_ctx == NULL) {
-		to_string_ctx = cypher_ast_annotation_context();
-		// set release handler to free allocated string
-		cypher_ast_annotation_context_set_release_handler(to_string_ctx, _free_to_string_annotation, NULL);
-		AST_AnnotationCtxCollection_SetToStringCtx(ast->anot_ctx_collection, to_string_ctx);
-	}
-
 	char * str = (char *)cypher_astnode_get_annotation(to_string_ctx, node);
 	if(str == NULL) {
-		struct cypher_input_range range = cypher_astnode_range(node);
-		uint length = range.end.offset - range.start.offset + 2;
-		str = rm_malloc(sizeof(char) * length);
-		strncpy(str, ctx->query_data.query + range.start.offset, length - 1);
-		str[length - 1] = '\0';
+		cypher_astnode_type_t t = cypher_astnode_type(node);
+		const cypher_astnode_t *ast_identifier = NULL;
+		if(t == CYPHER_AST_NODE_PATTERN) {
+			ast_identifier = cypher_ast_node_pattern_get_identifier(node);
+		} else if(t == CYPHER_AST_REL_PATTERN) {
+			ast_identifier = cypher_ast_rel_pattern_get_identifier(node);
+		} else {
+			struct cypher_input_range range = cypher_astnode_range(node);
+			uint length = range.end.offset - range.start.offset + 1;
+			str = rm_malloc(sizeof(char) * length);
+			strncpy(str, ctx->query_data.query_no_params + range.start.offset, length - 1);
+			str[length - 1] = '\0';
+		}
+		if(ast_identifier) {
+			// Graph entity has a user-defined alias return it.
+			return cypher_ast_identifier_get_name(ast_identifier);
```

Comment:
This casts the value from `const char *` to `char *`. I believe there would be no problem with this function returning a `const char *`.

Reply:
done